### PR TITLE
Only complete PipeReader/PipeWriter after operations complete

### LINKED ISSUE #350: HeaderDelimitedMessageHandler.Write throws when in race with disposal
If the `PipeMessageHandler` is writing (or reading) while being disposed, it can create a memory corruption and/or exception situation when the PipeWriter or PipeReader or their buffers are accessed after being completed.

VS has recently shown a tendency to fail with an infobar with this error:

```System.InvalidOperationException: Writing is not allowed after writer was completed.
 at System.IO.Pipelines.Pipe.GetMemory(Int32 sizeHint)
 at System.IO.Pipelines.Pipe.DefaultPipeWriter.GetMemory(Int32 sizeHint)
 at StreamJsonRpc.HeaderDelimitedMessageHandler.Write(JsonRpcMessage content, CancellationToken cancellationToken)
 at StreamJsonRpc.PipeMessageHandler.WriteCoreAsync(JsonRpcMessage content, CancellationToken cancellationToken)
 at StreamJsonRpc.MessageHandlerBase.<WriteAsync>d__15.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at StreamJsonRpc.JsonRpc.<TransmitAsync>d__118.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Threading.Tasks.ValueTask.ThrowIfCompletedUnsuccessfully()
 at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__102.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__96`1.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.VisualStudio.CodeLens.Proxy.CodeLensHubClient.<>c__DisplayClass22_0`1.<<InvokeRpcAsync>b__0>d.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.<CallExtensionPointAsync>d__36`1.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

### PR DESCRIPTION
**Pull request by AArnott**
We must not allow a race to exist between PipeReader/PipeWriter use and calling Complete() on them. This not only produces exceptions like the one in #350 but it opens up the possibility of accessing recycled buffers.

Fixes #350

### CODE REVIEW

File: src/StreamJsonRpc/MessageHandlerBase.cs [hunk 1]
```diff
@@ -35,6 +35,16 @@ public abstract class MessageHandlerBase : IJsonRpcMessageHandler, IDisposableOb
         /// </summary>
         private readonly AsyncSemaphore sendingSemaphore = new AsyncSemaphore(1);
 
+        /// <summary>
+        /// The sync object.
```
Comment (jepetty) [inline src/StreamJsonRpc/MessageHandlerBase.cs]: Small nit: could we add something about this locking/protecting specifically the state of the MessageHandler?

Comment (AArnott) [reply src/StreamJsonRpc/MessageHandlerBase.cs]: Good idea.

**Comment by davidfowl** [comment]
Why dispose in Read and Write itself instead of Dispose?

**Comment by AArnott** [comment]
@davidfowl We dispose in all 3 places (read, write, dispose), whichever comes *last*. The reason is because we auto-dispose when the connection drops, or whenever the user calls Dispose explicitly. But if we're in the middle of a write/read operation, that's the "use after dispose" case you warned me about and that we're seeing exceptions regarding. So we can't Complete during a read/write operation, and our Dispose is synchronous, so the best we can do is defer completion until after the read/write operations complete.

**Comment by davidfowl** [comment]
>  The reason is because we auto-dispose when the connection drops, or whenever the user calls Dispose explicitly.

Don't auto-dispose when the connection drops. Only when they call dispose, that's how you know they are done with the handler.

**Comment by AArnott** [comment]
> Don't auto-dispose when the connection drops. Only when they call dispose, that's how you know they are done with the handler.

This is prior shipped behavior, so it would be difficult to change. But anyway, the fact that they call `Dispose` does not guarantee that we're done reading, because they really have no way of knowing whether a message was coming in just as they decided to dispose.

**Comment by davidfowl** [comment]
> This is prior shipped behavior, so it would be difficult to change. 

What's the observed behavior from the caller's end? Can't that be preserved?

> But anyway, the fact that they call Dispose does not guarantee that we're done reading, because they really have no way of knowing whether a message was coming in just as they decided to dispose.

But that code would throw an object disposed exception, because you would cancel any pending operations, then dispose.

**Comment by AArnott** [comment]
We do have #246 to better define exactly what the disposal behaviors and expectations are.

From this completed PR we already have a failing test in the VS repo because the test verified that the underlying pipe had been closed immediately after JsonRpc.Dispose() returned. So I have to go back and fix that, at least. I'd rather not return from Dispose() before *all* async/concurrent code has stopped executing, since anything less can result in crashes when the AppDomain is torn down if that code is still active.

> But that code would throw an object disposed exception, because you would cancel any pending operations, then dispose.

It's not enough to cancel then dispose, I think. I have to wait for the canceled work to complete. Otherwise I might still be disposing the PipeReader/PipeWriter while it's being accessed on another thread but before that thread has observed the cancellation request.

**Comment by davidfowl** [comment]
Fair, this is the safest approach then (especially since there's no dispose async)

**Comment by AArnott** [comment]
Thanks for your review.

However this PR is faulty after all, as it blocked a VS insertion due to a failing test, since Dispose() returned before all the async cleanup was done. So I have a change nearly ready for another PR that (wait for it...) adds a DisposeAsync() method. Since you said "especially since there's no dispose async", I'm very interested in hearing your concerns about that approach. I'll include you on the PR when it's ready so you can see more specifically what I'm talking about.

**Pull request merged**